core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/WindowsShareCopy.java

```diff
@@ -0,0 +1,267 @@
+
+    // Authentication with NTLM and read the directory from the Windows Share.
+    final NtlmPasswordAuthentication auth = new NtlmPasswordAuthentication(config.netBiosDomainName,
+                                                                           config.netBiosUsername,
+                                                                           config.netBiosPassword);
+    SmbFile dir = new SmbFile(smbDirectory, auth);
+
+    // Create the HDFS directory if it doesn't exist.
+    final Path hdfsDir = new Path(config.destinationDirectory);
+    final FileSystem hdfs = hdfsDir.getFileSystem(new Configuration());
+    if (!hdfs.exists(hdfsDir)) {
+      hdfs.mkdirs(hdfsDir);
+    }
+
+    // Determines the buffer size to be used during writing.
+    Configuration conf = new Configuration();
+    conf.setLong("io.file.buffer.size", config.bufferSize);
+    String[] files = dir.list();
+
+    // Copies the files in a multithreaded way
+    CountDownLatch executorTerminateLatch = new CountDownLatch(1);
+    ExecutorService executorService = createExecutor(config.numThreads, executorTerminateLatch);
+    CompletionService<String> completionService = new ExecutorCompletionService<>(executorService);
+
+    try {
+      for (final String file : files) {
+        completionService.submit(new Callable<String>() {
+          @Override
+          public String call() throws Exception {
+            try {
+              if (smbDirectory.endsWith("/")) {
+                return copyFileToHDFS(hdfs, smbDirectory + file, hdfsDir, auth);
+              } else {
+                return copyFileToHDFS(hdfs, smbDirectory + "/" + file, hdfsDir, auth);
+              }
+            } catch (Exception e) {
+              LOG.warn("Exception while copying the file {}", file, e);
+              return null;
+            }
+          }
+        });
+      }
+
+      for (int i = 0; i < files.length; i++) {
+        try {
+          Future<String> fileWritten = completionService.take();
+          String fileName = fileWritten.get();
+          if (fileName != null) {
+            LOG.info("{} is copied", fileName);
+          }
+        } catch (Throwable t) {
+          Throwables.propagateIfPossible(t, Exception.class);
+          throw Throwables.propagate(t);
+        }
+      }
+    } finally {
+      executorService.shutdownNow();
+      executorTerminateLatch.await();
+    }
+  }
+
+  private String copyFileToHDFS(FileSystem hdfs, String smbSourceFile, Path dest, NtlmPasswordAuthentication auth)
+    throws IOException {
+    SmbFile smbFile = new SmbFile(smbSourceFile, auth);
+    String name = smbFile.getName();
+    Path destFile = new Path(dest, name);
+    LOG.debug("Thread {} is accessing source file {}, dest file {}", Thread.currentThread().getName(),
+              smbFile.getName(), destFile.getName());
+    // If file already exists, then we skip over.
+    if (hdfs.exists(destFile)) {
```

Comment:
Shouldn't you allow user to overwrite ? Take a configuration to allow overwrite or not. I think it will be a common use case where you want to run copy again and again with change in data at windows share. With this the use has to go to hdfs and delete and then run the pipeline.